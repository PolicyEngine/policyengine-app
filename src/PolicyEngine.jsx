import Home from "./pages/Home";
import Research from "./pages/Research";
import About from "./pages/About";
import Jobs from "./pages/Jobs";
import {
  Navigate,
  Outlet,
  Route,
  Routes,
  useSearchParams,
} from "react-router-dom";
import Donate from "./pages/Donate";
import { useLocation } from "react-router-dom";
import BlogPage from "./pages/BlogPage";
import { COUNTRY_BASELINE_POLICIES, COUNTRY_CODES } from "./data/countries";

import { useEffect, useState, lazy, Suspense } from "react";
import {
  apiCall,
  copySearchParams,
  countryApiCall,
  updateMetadata,
} from "./api/call";
import LoadingCentered from "./layout/LoadingCentered";
import ErrorPage from "./layout/ErrorPage";
import Header from "./layout/Header";
import Testimonials from "./pages/Testimonials";
import CalculatorInterstitial from "./pages/CalculatorInterstitial";
import CitizensEconomicCouncil from "./applets/CitizensEconomicCouncil";
import APIDocumentationPage from "./pages/APIDocumentationPage";
import CookieConsent from "./modals/CookieConsent";
import TrafwaCalculator from "./applets/TrafwaCalculator";
import StateEitcsCtcs from "./applets/StateEitcsCtcs";
import AuthCallback from "./layout/AuthCallback";
import UserProfilePage from "./pages/UserProfilePage";
import PrivacyPage from "./pages/PrivacyPage";
import TACPage from "./pages/TermsAndConditions";
import { useAuth0 } from "@auth0/auth0-react";
import { ConfigProvider } from "antd";
import style from "./style";
import RedirectToCountry from "./routing/RedirectToCountry";
import CountryIdLayout from "./routing/CountryIdLayout";
import RedirectBlogPost from "./routing/RedirectBlogPost";
import { StatusPage } from "./pages/StatusPage";
import ManifestosComparison from "./applets/ManifestosComparison";

const PolicyPage = lazy(() => import("./pages/PolicyPage"));
const HouseholdPage = lazy(() => import("./pages/HouseholdPage"));

function ScrollToTop() {
  const { pathname } = useLocation();

  useEffect(() => {
    window.scrollTo(0, 0);
  }, [pathname]);

  return null;
}

export default function PolicyEngine() {
  // This will either return the country ID or null,
  // but the page is guaranteed to redirect to an ID
  const countryId = extractCountryId();

  const [searchParams, setSearchParams] = useSearchParams();
  const householdId = searchParams.get("household");

  let defaultBaselinePolicy = 1;
  if (COUNTRY_CODES.includes(countryId)) {
    defaultBaselinePolicy = COUNTRY_BASELINE_POLICIES[countryId];
  }

  const reformPolicyId = searchParams.get("reform") || defaultBaselinePolicy;
  const baselinePolicyId =
    searchParams.get("baseline") || defaultBaselinePolicy;

  const [metadata, setMetadata] = useState(null);
  const [metadataError, setMetadataError] = useState(false);

  const [baselinePolicy, setBaselinePolicy] = useState({
    id: baselinePolicyId,
    label: null,
    data: null,
  });
  const [reformPolicy, setReformPolicy] = useState({
    id: reformPolicyId,
    label: null,
    data: null,
  });
  const policy = {
    baseline: baselinePolicy,
    reform: reformPolicy,
  };

  const [hasShownHouseholdPopup, setHasShownHouseholdPopup] = useState(false);
  const [userProfile, setUserProfile] = useState({});

  // Update the metadata state when something happens to
  // the countryId (e.g. the user changes the country).
  useEffect(() => {
    // If we're accessing the page without a country ID,
    // our router will handle redirecting to a country ID;
    // this process is guaranteed, thus we will just not fetch
    // in this situation
    if (!countryId) {
      return;
    }

    async function asyncUpdateMetadata() {
      try {
        const metadata = await updateMetadata(countryId);
        if (!metadata) {
          setMetadataError(true);
          setMetadata(null);
        } else {
          setMetadata(metadata);
          setMetadataError(false);
        }
      } catch (e) {
        console.error(e);
        setMetadataError(true);
        setMetadata(null);
      }
    }

    asyncUpdateMetadata();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [countryId]);

  // Get the baseline policy data when the baseline policy ID changes.
  useEffect(() => {
    if (metadata) {
      countryApiCall(countryId, `/policy/${baselinePolicyId}`)
        .then((res) => res.json())
        .then((dataHolder) => {
          if (dataHolder.result.label === "None") {
            dataHolder.result.label = null;
          }
          setBaselinePolicy({
            data: dataHolder.result.policy_json,
            label: dataHolder.result.label,
            id: baselinePolicyId,
          });
        });
    }
  }, [baselinePolicyId, countryId, metadata]);

  // Get the reform policy data when the reform policy ID changes.
  useEffect(() => {
    if (metadata) {
      countryApiCall(countryId, `/policy/${reformPolicyId}`)
        .then((res) => res.json())
        .then((dataHolder) => {
          if (dataHolder.result.label === "None") {
            dataHolder.result.label = null;
          }
          setReformPolicy({
            data: dataHolder.result.policy_json,
            label: dataHolder.result.label,
            id: reformPolicyId,
          });
        });
    }
  }, [countryId, reformPolicyId, metadata]);

  useEffect(() => {
    if (searchParams.get("renamed") && reformPolicyId) {
      countryApiCall(countryId, `/policy/${reformPolicyId}`)
        .then((res) => res.json())
        .then((dataHolder) => {
          setReformPolicy({
            data: dataHolder.result.policy_json,
            label: dataHolder.result.label,
            id: reformPolicyId,
          });
          let newSearch = copySearchParams(searchParams);
          newSearch.delete("renamed");
          setSearchParams(newSearch);
        });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [countryId, searchParams.get("renamed")]);

  // When metadata exists and a user is logged in, fetch user profile
  const { isAuthenticated, user } = useAuth0();
  useEffect(() => {
    async function fetchUserProfile() {
      const USER_PROFILE_PATH = `/${countryId}/user_profile`;
      // Determine if user already exists in user profile db
      try {
        const resGet = await apiCall(
          USER_PROFILE_PATH + `?auth0_id=${user.sub}`,
        );
        const resGetJson = await resGet.json();

        if (resGet.status === 200) {
          return resGetJson.result;
        } else if (resGet.status === 404 && resGetJson.status === "ok") {
          // If not, create user first, then fetch user
          const body = {
            auth0_id: user.sub,
            primary_country: countryId,
            user_since: Date.now(),
          };
          const resPost = await apiCall(USER_PROFILE_PATH, body, "POST");
          const resPostJson = await resPost.json();
          if (resPost.status !== 201) {
            console.error(
              `Error while trying to create new user with auth0_id ${user.sub}`,
            );
            console.error(resPostJson);
            return;
          }
          return resPostJson.result;
        } else {
          console.error(
            `Error while attempting to fetch user profile for user ${user.sub}`,
          );
          console.error(resGetJson);
          return;
        }
      } catch (err) {
        console.error(
          `Connection error while attempting to fetch user profile for user ${user.sub}: ${err}`,
        );
        return;
      }
    }

    if (countryId && isAuthenticated && user?.sub) {
      fetchUserProfile().then((userProfile) => setUserProfile(userProfile));
    }
  }, [countryId, user?.sub, isAuthenticated]);

  const loadingPage = (
    <>
      <Header />
      <LoadingCentered />
    </>
  );

  const householdPage = (
    <Suspense fallback={loadingPage}>
      <Header />
      <HouseholdPage
        metadata={metadata}
        householdId={householdId}
        policy={policy}
        hasShownHouseholdPopup={hasShownHouseholdPopup}
        setHasShownHouseholdPopup={setHasShownHouseholdPopup}
      />
    </Suspense>
  );

  const errorPage = <ErrorPage />;

  const policyPage = (
    <Suspense fallback={loadingPage}>
      <Header />
      <PolicyPage
        metadata={metadata}
        householdId={householdId}
        policy={policy}
        userProfile={userProfile}
      />
    </Suspense>
  );

  // If the path is /, redirect to /[countryId]
  // If the path is /[countryId], render the homepage
  // If the path is /[countryId]/about, render the about page
  // If the path is /[countryId]/jobs, render the jobs page
  // If the path is /[countryId]/research, render the research page
  // If the path is not recognized, redirect to /[countryId]

  return (
    <ConfigProvider
      theme={{
        token: {
          borderRadius: 0,
          colorPrimary: style.colors.BLUE,
          fontFamily: "Roboto Serif",
        },
      }}
    >
      <ScrollToTop />
      <CookieConsent />
      <Routes>
        {/* Redirect from / to /[countryId] */}
        <Route path="/" element={<RedirectToCountry />} />
        <Route path="/callback" element={<AuthCallback />} />
        <Route path="/:countryId" element={<CountryIdLayout />}>
          <Route index={true} element={<Home />} />
          <Route path="about" element={<About />} />
          <Route path="jobs" element={<Jobs />} />
          <Route path="testimonials" element={<Testimonials />} />
          <Route path="calculator" element={<CalculatorInterstitial />} />
          <Route path="research" element={<Outlet />}>
            <Route index={true} element={<Research />} />
            <Route path=":postName" element={<BlogPage />} />
          </Route>
          <Route path="donate" element={<Donate />} />
          <Route path="privacy" element={<PrivacyPage />} />
          <Route path="terms" element={<TACPage />} />

          <Route
            path="household/*"
            element={
              metadataError ? errorPage : metadata ? householdPage : loadingPage
            }
          />
          <Route
            path="policy/*"
            element={
              metadataError ? errorPage : metadata ? policyPage : loadingPage
            }
          />

          <Route
            path="profile"
            element={
              !userProfile ? (
                <ErrorPage />
              ) : (
                <Navigate to={`/${countryId}/profile/${userProfile.user_id}`} />
              )
            }
          />
          <Route
            path="profile/:user_id"
            element={
              <UserProfilePage
                metadata={metadata}
                authedUserProfile={userProfile}
                metadataError={metadataError}
              />
            }
          />

          <Route
            path="api"
            element={<APIDocumentationPage metadata={metadata} />}
          />
          {/* redirect from /countryId/blog/slug to /countryId/research/slug */}
          <Route path="blog/:postName" element={<RedirectBlogPost />} />
        </Route>
        <Route path="/uk/cec" element={<CitizensEconomicCouncil />} />
        <Route path="/uk/2024-manifestos" element={<ManifestosComparison />} />
        <Route path="/:countryId/api_status" element={<StatusPage />} />
        <Route
          path="/us/trafwa-ctc-calculator"
          element={<TrafwaCalculator />}
        />
        <Route path="/us/state-eitcs-ctcs" element={<StateEitcsCtcs />} />

        {/* Redirect for unrecognized paths */}
        <Route path="*" element={<Navigate to={`/${countryId}`} />} />
      </Routes>
    </ConfigProvider>
  );
}

/**
 * Extracts country ID from path or returns null if one doesn't exist
 * @returns {String|null}
 */
function extractCountryId() {
  const path = window.location.pathname;
  const pathParts = path.split("/").filter((item) => item.length > 0);

  // If valid, return the normal country ID
  if (pathParts.length > 0 && COUNTRY_CODES.includes(pathParts[0])) {
    return pathParts[0];
  }
  // If we have an invalid ID (e.g., "undefined" or "garbage"),
  // the router will redirect to "us", so return that as country ID
  else if (pathParts.length > 0) {
    return "us";
  }
  // Otherwise, we're on the standard page; return null and allow
  // router to redirect
  else {
    return null;
  }
}
