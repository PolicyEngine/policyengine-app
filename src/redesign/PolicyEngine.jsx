import { useState } from "react";
import { IntroductionToPolicyEngine } from "./components/content/IntroductionToPolicyEngine";
import { HeaderBar } from "./components/content/Header";
import { CalculatorCallToAction } from "./components/content/CalculatorCallToAction";
import { SubscribeToPolicyEngine } from "./components/content/SubscribeToPolicyEngine";
import { BlogPreview } from "./components/content/BlogPreview";

export default function PolicyEngine() {
  const [countryId] = useState("us");
  return (
    <div>
      <HeaderBar />
      <IntroductionToPolicyEngine countryId={countryId} />
      <CalculatorCallToAction />
      <BlogPreview />
      <SubscribeToPolicyEngine />
    </div>
  );
}
