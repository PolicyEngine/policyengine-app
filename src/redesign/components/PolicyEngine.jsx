import Home from "./Home";
import Research from "./Research";
import About from "./About";
import Jobs from "./Jobs";
import { Navigate, Route, Routes, useSearchParams } from "react-router-dom";
import Contact from "./Contact";
import Donate from "./Donate";
import { useLocation } from "react-router-dom";
import BlogPage from "../../pages/BlogPage";

import { useEffect, useState, lazy, Suspense } from "react";
import {
  copySearchParams,
  countryApiCall,
  updateMetadata,
} from "../../api/call";
import LoadingCentered from "../../layout/LoadingCentered";
import ErrorPage from "../../layout/Error";
import Header from "./Header";
import Testimonials from "./Testimonials";
import CalculatorInterstitial from "./CalculatorInterstitial";
import CitizensEconomicCouncil from "./CitizensEconomicCouncil";
import APIDocumentationPage from "../../pages/APIDocumentationPage";
import CookieConsent from "layout/CookieConsent";
import TrafwaCalculator from "./TrafwaCalculator";
import PrivacyPage from "../../pages/PrivacyPage";
import TACPage from "../../pages/TermsAndConditions";

const PolicyPage = lazy(() => import("../../pages/PolicyPage"));
const HouseholdPage = lazy(() => import("../../pages/HouseholdPage"));

function ScrollToTop() {
  const { pathname } = useLocation();

  useEffect(() => {
    window.scrollTo(0, 0);
  }, [pathname]);

  return null;
}

export default function PolicyEngine({ pathname }) {
  const COUNTRIES = ["us", "uk", "ca", "ng", "il"];

  // First, check if the country is specified (.org/[country]/...)
  const path = pathname || window.location.pathname;
  const pathParts = path.split("/");
  let countryId = pathParts[1];

  if (!COUNTRIES.includes(countryId)) {
    // If the country is not specified, look up the country ID from the user's browser language
    const browserLanguage = navigator.language;
    countryId = {
      "en-US": "us",
      "en-GB": "uk",
      "en-CA": "ca",
      "en-NG": "ng",
      "en-IL": "il",
    }[browserLanguage];
  }

  const [searchParams, setSearchParams] = useSearchParams();
  const householdId = searchParams.get("household");
  const defaultBaselinePolicy =
    countryId === "uk"
      ? 1
      : countryId === "us"
        ? 2
        : countryId === "ca"
          ? 3
          : countryId === "ng"
            ? 4
            : countryId === "il"
              ? 5
              : 1;
  const reformPolicyId = searchParams.get("reform") || defaultBaselinePolicy;
  const baselinePolicyId =
    searchParams.get("baseline") || defaultBaselinePolicy;

  const [metadata, setMetadata] = useState(null);
  const [error] = useState(null);

  const [baselinePolicy, setBaselinePolicy] = useState({
    id: baselinePolicyId,
    label: null,
    data: null,
  });
  const [reformPolicy, setReformPolicy] = useState({
    id: reformPolicyId,
    label: null,
    data: null,
  });
  const policy = {
    baseline: baselinePolicy,
    reform: reformPolicy,
  };

  const [hasShownHouseholdPopup, setHasShownHouseholdPopup] = useState(false);
  const [hasShownPopulationImpactPopup, setHasShownPopulationImpactPopup] =
    useState(false);

  // Update the metadata state when something happens to the countryId (e.g. the user changes the country).
  useEffect(() => {
    try {
      updateMetadata(countryId, setMetadata);
    } catch (e) {
      // Sometimes this fails. When it does, refresh the page, but only once (use a param in the URL to make sure it only happens once).
      if (!searchParams.get("refreshed")) {
        let newSearch = copySearchParams(searchParams);
        newSearch.set("refreshed", true);
        setSearchParams(newSearch);
        window.location.reload();
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [countryId]);

  // Get the baseline policy data when the baseline policy ID changes.
  useEffect(() => {
    if (metadata) {
      countryApiCall(countryId, `/policy/${baselinePolicyId}`)
        .then((res) => res.json())
        .then((dataHolder) => {
          if (dataHolder.result.label === "None") {
            dataHolder.result.label = null;
          }
          setBaselinePolicy({
            data: dataHolder.result.policy_json,
            label: dataHolder.result.label,
            id: baselinePolicyId,
          });
        });
    }
  }, [baselinePolicyId, countryId, metadata]);

  // Get the reform policy data when the reform policy ID changes.
  useEffect(() => {
    if (metadata) {
      countryApiCall(countryId, `/policy/${reformPolicyId}`)
        .then((res) => res.json())
        .then((dataHolder) => {
          if (dataHolder.result.label === "None") {
            dataHolder.result.label = null;
          }
          setReformPolicy({
            data: dataHolder.result.policy_json,
            label: dataHolder.result.label,
            id: reformPolicyId,
          });
        });
    }
  }, [countryId, reformPolicyId, metadata]);

  useEffect(() => {
    if (searchParams.get("renamed") && reformPolicyId) {
      countryApiCall(countryId, `/policy/${reformPolicyId}`)
        .then((res) => res.json())
        .then((dataHolder) => {
          setReformPolicy({
            data: dataHolder.result.policy_json,
            label: dataHolder.result.label,
            id: reformPolicyId,
          });
          let newSearch = copySearchParams(searchParams);
          newSearch.delete("renamed");
          setSearchParams(newSearch);
        });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [countryId, searchParams.get("renamed")]);

  const loadingPage = <LoadingCentered />;

  const householdPage = (
    <Suspense fallback={loadingPage}>
      <Header />
      <HouseholdPage
        metadata={metadata}
        householdId={householdId}
        policy={policy}
        hasShownHouseholdPopup={hasShownHouseholdPopup}
        setHasShownHouseholdPopup={setHasShownHouseholdPopup}
      />
    </Suspense>
  );

  const errorPage = <ErrorPage />;

  const policyPage = (
    <Suspense fallback={loadingPage}>
      <Header />
      <PolicyPage
        metadata={metadata}
        householdId={householdId}
        policy={policy}
        hasShownPopulationImpactPopup={hasShownPopulationImpactPopup}
        setHasShownPopulationImpactPopup={setHasShownPopulationImpactPopup}
      />
    </Suspense>
  );

  // If the path is /, redirect to /[countryId]
  // If the path is /[countryId], render the homepage
  // If the path is /[countryId]/about, render the about page
  // If the path is /[countryId]/jobs, render the jobs page
  // If the path is /[countryId]/research, render the research page
  // If the path is not recognized, redirect to /[countryId]

  return (
    <>
      <ScrollToTop />
      <CookieConsent />
      <Routes>
        {/* Redirect from / to /[countryId] */}
        <Route path="/" element={<Navigate to={`/${countryId}`} />} />

        <Route path="/:countryId" element={<Home />} />
        <Route path="/:countryId/about" element={<About />} />
        <Route path="/:countryId/jobs" element={<Jobs />} />
        <Route path="/:countryId/testimonials" element={<Testimonials />} />
        <Route
          path="/:countryId/calculator"
          element={<CalculatorInterstitial />}
        />
        <Route path="/:countryId/research" element={<Research />} />
        <Route path="/:countryId/contact" element={<Contact />} />
        <Route path="/:countryId/donate" element={<Donate />} />
        <Route path="/:countryId/research/*" element={<BlogPage />} />
        <Route path="/:countryId/privacy" element={<PrivacyPage />} />
        <Route path="/:countryId/terms" element={<TACPage />} />

        <Route
          path="/:countryId/household/*"
          element={metadata ? householdPage : error ? errorPage : loadingPage}
        />
        <Route
          path="/:countryId/policy/*"
          element={metadata ? policyPage : error ? errorPage : loadingPage}
        />

        <Route
          path="/:countryId/api"
          element={<APIDocumentationPage metadata={metadata} />}
        />
        <Route path="/uk/cec" element={<CitizensEconomicCouncil />} />
        <Route
          path="/us/trafwa-ctc-calculator"
          element={<TrafwaCalculator />}
        />

        {/* redirect from /countryId/blog/slug to /countryId/research/slug */}
        <Route
          path="/:countryId/blog/:slug"
          element={<Navigate to={`/${countryId}/research/${pathParts[3]}`} />}
        />

        {/* Redirect for unrecognized paths */}
        <Route path="*" element={<Navigate to={`/${countryId}`} />} />
      </Routes>
    </>
  );
}
