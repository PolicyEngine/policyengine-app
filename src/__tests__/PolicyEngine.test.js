import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { BrowserRouter, useSearchParams } from "react-router-dom";
import postJson from "../posts/posts.json";

import PolicyEngine from "../PolicyEngine";

jest.mock("react-router-dom", () => {
  const originalModule = jest.requireActual("react-router-dom");
  return {
    __esModule: true,
    ...originalModule,
    useSearchParams: jest.fn(),
  };
});

jest.mock("react-plotly.js", () => jest.fn());

jest.mock("react-router-dom", () => {
  const originalModule = jest.requireActual("react-router-dom");
  return {
    __esModule: true,
    ...originalModule,
    useSearchParams: jest.fn(),
  };
});

const { location } = window;

beforeAll(async () => {
  window.scrollTo = jest.fn();

  delete window.location;
  window.location = {
    reload: jest.fn(),
  };
});

afterAll(() => {
  window.location = location;
  jest.resetAllMocks();
});

describe("Test main PolicyEngine component", () => {
  test("Renders for US if proper data passed", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/us",
      origin: "https://www.policyengine.org/us",
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(
      getByText("Computing Public Policy for Everyone"),
    ).toBeInTheDocument();
    expect(getByText("Trusted across the US")).toBeInTheDocument();
  });
  test("Renders for UK if proper data passed", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/uk",
      origin: "https://www.policyengine.org/uk",
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(
      getByText("Computing Public Policy for Everyone"),
    ).toBeInTheDocument();
    expect(getByText("Trusted across the UK")).toBeInTheDocument();
  });
  test("Routes to auth callback", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/callback",
      origin: "https://www.policyengine.org/callback",
    };

    const { queryByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(queryByText("Follow us on social media")).not.toBeInTheDocument();
  });
  test("Routes to about page for US", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/us/about",
      origin: "https://www.policyengine.org/us/about",
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(getByText("Our people")).toBeInTheDocument();
  });
  test("Routes to jobs page for UK", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/uk/jobs",
      origin: "https://www.policyengine.org/uk/jobs",
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(getByText("Join Our Team")).toBeInTheDocument();
  });
  test("Routes to testimonials page for US", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/us/testimonials",
      origin: "https://www.policyengine.org/us/testimonials",
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(getByText("What people say about PolicyEngine")).toBeInTheDocument();
  });
  test("Routes to calculator interstitial page for UK", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/uk/calculator",
      origin: "https://www.policyengine.org/uk/calculator",
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(getByText("Choose a calculator")).toBeInTheDocument();
  });
  test("Routes to research page for US", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/us/research",
      origin: "https://www.policyengine.org/us/research",
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(getByText("Research and analysis")).toBeInTheDocument();
    // The below test should be re-enabled when tags on the page are no longer text nodes
    // inside divs, which does not follow proper semantic HTML and cannot be tested using getByText
    // expect(getByText("us")).toBeInTheDocument();
  });
  test("Routes to research page for UK", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/uk/research",
      origin: "https://www.policyengine.org/uk/research",
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(getByText("Research and analysis")).toBeInTheDocument();
    // The below test should be re-enabled when tags on the page are no longer text nodes
    // inside divs, which does not follow proper semantic HTML and cannot be tested using getByText
    // expect(getByText("uk")).toBeInTheDocument();
  });
  test("Routes to donate page for UK", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/uk/donate",
      origin: "https://www.policyengine.org/uk/donate",
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(getByText("The Difference Your Support Makes")).toBeInTheDocument();
  });

  // This test only ensures that the JS page renders, not that markdown is transformed,
  // as markdown files are stubbed out in our current implementation of Jest.
  // Processing markdown properly, which should be done to test blog files,
  // would likely require a custom webpack setup.
  test("Routes to individual blog posts for US", () => {
    global.fetch = jest.fn().mockImplementationOnce(() =>
      Promise.resolve({
        text: () => Promise.resolve("success"),
      }),
    );

    // Filter posts.json for US articles
    const filteredPostJson = postJson.filter((post) =>
      post.tags.includes("us"),
    );

    // Choose one at random
    const randIndex = Math.floor(Math.random() * filteredPostJson.length);

    const selectedPost = filteredPostJson[randIndex];
    const postFilepath = selectedPost.filename.split(".")[0];

    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: `/us/research/${postFilepath}`,
      origin: `https://www.policyengine.org/us/research/${postFilepath}`,
    };

    render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    const link = screen.getByRole("link", { name: /Research/i });
    expect(link).toBeInTheDocument();
  });
  test("Routes to privacy page for UK", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/uk/privacy",
      origin: "https://www.policyengine.org/uk/privacy",
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(getByText("Privacy")).toBeInTheDocument();
  });
  test("Routes to T&C page for US", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/us/terms",
      origin: "https://www.policyengine.org/us/terms",
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(getByText("Terms of Service")).toBeInTheDocument();
  });
  // Note: This test only determines if routing occurs;
  // more detailed testing should be done for the component
  // itself
  test("Routes to household page for US", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/us/household",
      origin: "https://www.policyengine.org/us/household",
    };

    const { queryByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(
      queryByText("Computing Public Policy for Everyone"),
    ).not.toBeInTheDocument();
  });
  // Note: This test only determines if routing occurs;
  // more detailed testing should be done for the component
  // itself
  test("Routes to policy page for UK", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/uk/policy",
      origin: "https://www.policyengine.org/uk/policy",
    };

    const { queryByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(
      queryByText("Computing Public Policy for Everyone"),
    ).not.toBeInTheDocument();
  });
  test("Routes to user profile page if supplied with a user ID for UK", () => {
    const testUserId = "1";

    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: `/uk/profile/${testUserId}`,
      origin: `https://www.policyengine.org/uk/profile/${testUserId}`,
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(getByText("Profile")).toBeInTheDocument();
  });
  test("Routes to API documentation page for US", () => {
    // This test uses node-fetch to actually fetch country metadata,
    // but can't use fetch to fetch the API component's sample data,
    // and since it's not possible to conditionally polyfill fetch,
    // this is a better way of silencing the inevitable 404 error
    jest.spyOn(console, "error").mockImplementation(() => {});

    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/us/api",
      origin: "https://www.policyengine.org/us/api",
    };

    const { getByText } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(getByText("PolicyEngine API Documentation")).toBeInTheDocument();
  });
  test("Routes to TRAFWA calculator", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/us/trafwa-ctc-calculator",
      origin: "https://www.policyengine.org/us/trafwa-ctc-calculator",
    };

    const { getByTitle } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(
      getByTitle("TRAFWA Child Tax Credit Calculator"),
    ).toBeInTheDocument();
  });
  test("Routes to Citizens Economic Council page", () => {
    useSearchParams.mockImplementation(() => {
      return [new URLSearchParams(), jest.fn()];
    });

    window.location = {
      ...window.location,
      pathname: "/uk/cec",
      origin: "https://www.policyengine.org/uk/cec",
    };

    const { getByTitle } = render(
      <BrowserRouter>
        <PolicyEngine />
      </BrowserRouter>,
    );

    expect(
      getByTitle("Citizens' Economic Council reform simulator"),
    ).toBeInTheDocument();
  });
});
