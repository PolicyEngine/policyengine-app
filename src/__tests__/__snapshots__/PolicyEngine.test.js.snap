// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test routing > should redirect an invalid country to the uk page 1`] = `"We compute the impact of public policy."`;

exports[`test routing > should redirect to the 404 page if bad url on a country page 1`] = `"Sorry, we can't find the page you're looking for."`;
